### DISCUSSION TITLE
Warm-starting from previous solutions

### DISCUSSION THREAD

**Discussion by johege** [PyVRP/PyVRP#710] [Q&A] [answered]
First of all, thank you very much for this package. It is very easy to understand and great to dive into vehicle routing.

I wanted to ask if it is possible to use PyVRP in a dynamic way.
Let's say I have found a solution to a problem. Now, if a small number of new clients are added to the problem data, I was hoping that I could use the previous solution, or the pool of solutions at stopping time of the search, to quickly find solutions for the new, slightly customized problem. Is this somehow possible?

**Comment by N-Wouda** [comment]
Hi @johege! 👋 

We do not support this directly, unfortunately. But it should not be hard to do, following [the example](https://pyvrp.org/examples/using_pyvrp_components.html#The-solve-function) on how to write your own `solve()` function. In particular, the following slightly condensed example should be able to do what you want:
```python
def solve(data, prev_best=None):
    rng = RandomNumberGenerator(seed=42)
    pm = PenaltyManager.init_from(data)
    pop = Population(broken_pairs_distance)

    neighbours = compute_neighbours(data)
    ls = LocalSearch(INSTANCE, rng, neighbours)

    for node_op in NODE_OPERATORS:
        ls.add_node_operator(node_op(data))

    for route_op in ROUTE_OPERATORS:
        ls.add_route_operator(route_op(data))

    init = [Solution.make_random(data, rng) for _ in range(25)]
    if prev_best:
        init.append(prev_best)

    algo = GeneticAlgorithm(data, pm, rng, pop, ls, srex, init)

    return algo.run(stop=MaxRuntime(10))
```

Suppose we have some initial data object, `data0`, and we call
```python
res0 = solve(data0)
```
then add some clients to obtain `data1`, and call
```python
res1 = solve(data1, res0.best)
```
This should work. With new clients, `res0.best` will no longer be feasible for `data1`, but that should not pose significant problems if we let the solver run for a few iterations. You might also want to store and (re)store the current penalties from the `PenaltyManager` between calls to `solve`, but that is easy enough to add yourself based on the example above.

I hope this helps. Please let us know if you have any further questions!

**Reply by N-Wouda** [reply]
Warm-starting is now tracked in #855 to be a normally supported feature, rather than something that requires a workaround like this.

**Comment by johege** [comment]
Hi @N-Wouda,

Thanks for the quick reply. It's really cool how quickly you respond to the issues and discussions in this repo. 😄
It took me some time to try it out. I had to make some adjustments because the line
```python
if prev_best:
        init.append(prev_best)
```
added the original best solution to the `init` solutions without changing the feasibility status of the object to infeasible. So the old solution (ignoring the new clients) was kept as the best feasible solution for the updated problem without including the new clients.
Even when only changing the status to infeasible and appending it to the `init` solutions, the `prev_best` solution was always kept as the best infeasible solution, because the additional travel time caused by the additional clients was not included there.
I created a new `Solution` object with all the new clients combined into a single new route and appended it to the routes of the `prev_best.routes()` and added that to the `init` solutions.

Here's the adapted code. Maybe it helps someone else as well 😄 

```python
def solve(new_data, prev_data=None, prev_best=None):
    rng = RandomNumberGenerator(seed=42)
    pm = PenaltyManager.init_from(new_data)
    pop = Population(broken_pairs_distance)

    neighbours = compute_neighbours(new_data)
    ls = LocalSearch(new_data, rng, neighbours)

    for node_op in NODE_OPERATORS:
        ls.add_node_operator(node_op(new_data))

    for route_op in ROUTE_OPERATORS:
        ls.add_route_operator(route_op(new_data))

    init = [Solution.make_random(new_data, rng) for _ in range(25)]
    if prev_best:
        num_old_clients = prev_data.num_clients + prev_data.num_depots
        num_new_clients = new_data.num_clients + new_data.num_depots
        new_clients_ids = list(range(num_old_clients, num_new_clients))
        new_route = Route(
            data=new_data,
            visits=new_clients_ids,
            vehicle_type=0
        )

        new_routes = prev_best.routes()
        new_routes.append(new_route)
        
        adapted_prev_best = Solution(new_data, new_routes)
        init.append(adapted_prev_best)

    algo = GeneticAlgorithm(new_data, pm, rng, pop, ls, srex, init)

    return algo.run(stop=MaxRuntime(10), display=True)
```

There are probably smarter ways to integrate the new clients into the previous routes. But this works for now.

**Reply by N-Wouda** [reply]
That's a nice solution! Looking at it again, I think the local search can also do these insertions for you, as well as a bit of further optimisation with the newly adjusted data. That should reduce the code a bit:
```python
if prev_best:
    adapted_prev_best = ls(prev_best, pm.cost_evaluator())
    init.append(adapted_prev_best)
```

**Reply by johege** [reply]
Ah yes. That seems to work as well at least on a small example. Thanks again.

When using this approach, it seems that the inserted solution is taken as the Feas. best at the beginning. But it takes about the same time for the Feas. avg. to reach similar values as if the prev. best solution wasn't used.

![image](https://github.com/user-attachments/assets/d0944e63-3de8-4dd9-a03f-6fbd707ef0a4)

![image](https://github.com/user-attachments/assets/0b672ad1-443c-46ec-a501-71177eb65faa)

Both approaches seem to take about the same amount of time to converge to a near-best value in this example.
(The example was a 500-client MDVRP with 3 depots, to which 10 new clients were added. In both cases, the MaxRuntime was set to 1 minute).
So I'm not sure that this is really an improvement over just computing a new solution from scratch.

**Reply by N-Wouda** [reply]
That's one of the reasons we don't spend a lot of time improving our initial pool of solutions: it rarely matters whether we started from good or random solutions after a few hundred iterations.

In your case, since you add only a few clients compared to the overall problem size, perhaps you can restart from the population at the end of your previous run? Running the local search over each of the feasible and infeasible solutions in the old population before inserting them into the new one would essentially warm-start the entire algorithm. But whether that makes much of a difference I cannot say.

**Reply by johege** [reply, marked as answer]
That's right. I thought about using the whole previous population. I just wasn't familiar enough with the code yet to do that.
I think I have that implemented now:

```python
def solve(new_data, prev_pop=None, max_runtime=60, seed=42):
    rng = RandomNumberGenerator(seed=seed)
    pm = PenaltyManager.init_from(new_data)
    pop = Population(broken_pairs_distance)

    neighbours = compute_neighbours(new_data)
    ls = LocalSearch(new_data, rng, neighbours)

    for node_op in NODE_OPERATORS:
        ls.add_node_operator(node_op(new_data))

    for route_op in ROUTE_OPERATORS:
        ls.add_route_operator(route_op(new_data))

    if prev_pop:
        init = []
        for sol in prev_pop:
            adapted_prev_sol = ls(sol, pm.cost_evaluator())
            init.append(adapted_prev_sol)
    else:
        init = [Solution.make_random(new_data, rng) for _ in range(25)]

    algo = GeneticAlgorithm(new_data, pm, rng, pop, ls, srex, init)

    return algo.run(stop=MaxRuntime(max_runtime), display=True), algo._pop
```

In my small example, this actually seems to bring some improvement. The warm-start variant seems to converge to a good solution faster within the first couple of iterations than starting from scratch. (I gave the calculations a higher MaxRuntime this time)

![image](https://github.com/user-attachments/assets/d807a1f1-052a-4ed1-9e59-99c8f4f03ff0)

I only tested this with one dataset and a random seed. So this may not be generalizable. But it seems promising for my case, if you want to get good solutions for a slightly adapted problem relatively fast.

**Reply by N-Wouda** [reply]
That looks correct to me, yes. Nice to see it speed up convergence quite a bit! If you want, a small improvement could be to just return `pop`, rather than `algo._pop` - they're the same object, so there should be no need to access the GA's private state.

Thanks for sharing!

**Comment by N-Wouda** [comment]
Hi @johege, I'm just quickly commenting to say that we now support this in PyVRP 0.13. You can pass an initial solution to `Model.solve()` or `solve()` by setting the `initial_solution` parameter. Hope that helps!
